schema: skip flaky TestAlterTableDMLInjection/drop_partial_index

```
No pass/skip/fail event found for test
=== RUN   TestAlterTableDMLInjection/drop_partial_index
=== PAUSE TestAlterTableDMLInjection/drop_partial_index
=== CONT  TestAlterTableDMLInjection/drop_partial_index
    dml_injection_test.go:500:
          Error Trace:  pkg/sql/schemachanger/schemachanger_test_test/pkg/sql/schemachanger/dml_injection_test.go:500
                              github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scrun/scrun.go:189
                              github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scrun/scrun.go:139
                              github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scdeps/run_deps.go:144
                              github.com/cockroachdb/cockroach/pkg/sql/internal.go:1670
                              github.com/cockroachdb/cockroach/pkg/sql/internal.go:1759
                              github.com/cockroachdb/cockroach/pkg/kv/txn.go:1012
                              github.com/cockroachdb/cockroach/pkg/kv/db.go:1055
                              github.com/cockroachdb/cockroach/pkg/kv/db.go:1018
                              github.com/cockroachdb/cockroach/pkg/kv/db.go:993
                              github.com/cockroachdb/cockroach/pkg/sql/internal.go:1746
                              github.com/cockroachdb/cockroach/pkg/sql/internal.go:1668
                              github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scdeps/run_deps.go:109
                              github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scrun/scrun.go:135
                              github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scjob/job.go:118
                              github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scjob/job.go:45
                              github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1668
                              github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1669
                              github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:477
                              github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:389
                              github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484
                              src/runtime/asm_amd64.s:1598
          Error:        [][]string{[]string{"PreCommitPhase:1", "PostCommitNonRevertiblePhase:1", "delete", "1"}, []string{"PreCommitPhase:1", "PostCommitNonRevertiblePhase:1", "update", "1"}, []string{"PreCommitPhase:1", "PostCommitNonRevertiblePhase:2", "delete", "1"}, []string{"PreCommitPhase:1", "PostCommitNonRevertiblePhase:2", "update", "1"}, []string{"PreCommitPhase:1", "PreCommitPhase:2", "update", "2"}, []string{"PreCommitPhase:1", "n/a", "n/a", "1"}, []string{"PreCommitPhase:2", "PostCommitNonRevertiblePhase:1", "delete", "1"}, []string{"PreCommitPhase:2", "PostCommitNonRevertiblePhase:1", "update", "1"}, []string{"PreCommitPhase:2", "PostCommitNonRevertiblePhase:2", "delete", "1"}, []string{"PreCommitPhase:2", "PostCommitNonRevertiblePhase:2", "update", "1"}, []string{"PreCommitPhase:2", "n/a", "n/a", "1"}, []string{"StatementPhase:1", "PostCommitNonRevertiblePhase:1", "delete", "1"}, []string{"StatementPhase:1", "PostCommitNonRevertiblePhase:1", "update", "1"}, []string{"StatementPhase:1", "PostCommitNonRevertiblePhase:2", "delete", "1"}, []string{"StatementPhase:1", "PostCommitNonRevertiblePhase:2", "update", "1"}, []string{"StatementPhase:1", "PreCommitPhase:1", "update", "2"}, []string{"StatementPhase:1", "PreCommitPhase:2", "update", "2"}, []string{"StatementPhase:1", "n/a", "n/a", "1"}, []string{"pre-schema-change", "n/a", "n/a", "1"}} does not contain []string{"PostCommitNonRevertiblePhase:1", "PostCommitNonRevertiblePhase:2", "delete", "1"}
          Test:         TestAlterTableDMLInjection/drop_partial_index
          Messages:     phaseOrdinal=PostCommitNonRevertiblePhase:1
panic: test timed out after 14m55s
```
Full build log: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelEssentialCi/12118044?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true&showLog=12118023_193406_590.193239.193404&logFilter=debug&logView=flowAware

--- a/pkg/sql/schemachanger/dml_injection_test.go
+++ b/pkg/sql/schemachanger/dml_injection_test.go
@@ -26,7 +26,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
-	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/exp/slices"
 )
@@ -303,7 +302,6 @@ func TestAlterTableDMLInjection(t *testing.T) {
 		{
 			desc:         "create index",
 			schemaChange: "CREATE INDEX idx ON tbl (val)",
-			skipIssue:    112421,
 		},
 		{
 			desc:         "drop index",
@@ -362,7 +360,6 @@ func TestAlterTableDMLInjection(t *testing.T) {
 			desc:         "drop partial index",
 			setup:        []string{"CREATE INDEX idx ON tbl (val) WHERE val > 1"},
 			schemaChange: "DROP INDEX idx",
-			skipIssue:    112417,
 		},
 		{
 			desc: "drop column with partial index",
@@ -402,7 +399,6 @@ func TestAlterTableDMLInjection(t *testing.T) {
 				"CREATE MATERIALIZED VIEW mv AS SELECT * FROM tbl@idx",
 			},
 			schemaChange: "DROP INDEX idx CASCADE",
-			skipIssue:    112418,
 		},
 	}
 
@@ -423,6 +419,7 @@ func TestAlterTableDMLInjection(t *testing.T) {
 			var sqlDB *sqlutils.SQLRunner
 			var clusterCreated atomic.Bool
 			poMap := make(map[phaseOrdinal]int)
+			poCompleted := make(map[phaseOrdinal]struct{})
 			var poSlice []phaseOrdinal
 			testCluster := serverutils.StartCluster(t, 1, base.TestClusterArgs{
 				ServerArgs: base.TestServerArgs{
@@ -442,17 +439,19 @@ func TestAlterTableDMLInjection(t *testing.T) {
 								}
 
 								if t.Failed() {
-									return errors.New("terminating schema change due to test failure")
+									// Just bail out of this hook if the test has failed,
+									// returning errors will break rollbacks.
+									return nil
 								}
 
 								currentStage := p.Stages[stageIdx]
 								currentPO := toPhaseOrdinal(currentStage)
 								errorMessage := fmt.Sprintf("phaseOrdinal=%s", currentPO)
 
-								// Capture all stages in the StatementPhase before they disappear.
-								if currentStage.Phase == scop.StatementPhase {
-									require.Empty(t, poMap, errorMessage)
-									require.Empty(t, poSlice, errorMessage)
+								// Capture all stages in the StatementPhase before they disappear,
+								// only if they haven't been collected (we could encounter retries).
+								if currentStage.Phase == scop.StatementPhase &&
+									len(poSlice) == 0 {
 									for i, s := range p.Stages {
 										po := toPhaseOrdinal(s)
 										poMap[po] = i
@@ -499,6 +498,11 @@ func TestAlterTableDMLInjection(t *testing.T) {
 									panic(fmt.Sprintf("slice contains duplicate elements a=%s b=%s %s", a, b, errorMessage))
 								})
 								actualResults := sqlDB.QueryStr(t, `SELECT 	insert_phase_ordinal, operation_phase_ordinal, operation, val FROM tbl`)
+								// Transaction retry errors can occur, so don't repeat the same
+								// DML if hit such a case to avoid flaky tests.
+								if _, exists := poCompleted[currentPO]; exists {
+									return nil
+								}
 								// Use subset instead of equals for better error output.
 								require.Subset(t, expectedResults, actualResults, errorMessage)
 								require.Subset(t, actualResults, expectedResults, errorMessage)
@@ -532,6 +536,7 @@ func TestAlterTableDMLInjection(t *testing.T) {
 									sqlDB.ExecWithMessage(t, errorMessage, insert, toAnySlice(currentPO.deleteRow(poSlice[i]))...)
 									sqlDB.ExecWithMessage(t, errorMessage, insert, toAnySlice(currentPO.updateRow(poSlice[i], false))...)
 								}
+								poCompleted[currentPO] = struct{}{}
 								return nil
 							},
 						},

